src/common/datasource/src/buffered_writer.rs

```diff
@@ -0,0 +1,142 @@
+pub trait ArrowWriterCloser {
+    async fn close(mut self) -> Result<FileMetaData>;
+}
+
+pub type DefaultBufferedWriter<E> = BufferedWriter<Compat<Writer>, E>;
+
+impl<T: AsyncWrite + Send + Unpin, U: DfRecordBatchEncoder + ArrowWriterCloser>
+    BufferedWriter<T, U>
+{
+    pub async fn close_with_arrow_writer(mut self) -> Result<(FileMetaData, u64)> {
+        if let Some(encoder) = self.encoder.take() {
+            let metadata = encoder.close().await?;
+            let written = self.try_flush(true).await?;
+
+            // It's important to shut down! flushes all pending writes
+            self.close().await?;
+            return Ok((metadata, written));
+        }
+        error::BufferedWriterClosedSnafu {}.fail()
+    }
+}
+
+impl<T: AsyncWrite + Send + Unpin, U: DfRecordBatchEncoder> BufferedWriter<T, U> {
+    pub async fn close(&mut self) -> Result<()> {
+        self.writer.shutdown().await.context(error::AsyncWriteSnafu)
+    }
+
+    pub fn new(threshold: usize, buffer: SharedBuffer, encoder: U, writer: T) -> Self {
+        Self {
+            threshold,
+            writer,
+            encoder: Some(encoder),
+            buffer,
+            bytes_written: 0,
+            flushed: false,
+        }
+    }
+
+    pub fn bytes_written(&self) -> u64 {
+        self.bytes_written
+    }
+
+    pub async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        if let Some(encoder) = &mut self.encoder {
+            encoder.write(batch)?;
+
+            self.try_flush(false).await?;
+
+            return Ok(());
+        }
+        error::BufferedWriterClosedSnafu {}.fail()
+    }
+
+    pub fn flushed(&self) -> bool {
+        self.flushed
+    }
+
+    pub async fn try_flush(&mut self, all: bool) -> Result<u64> {
+        let mut bytes_written: u64 = 0;
+
+        // Once buffered data size reaches threshold, split the data in chunks (typically 4MB)
+        // and write to underlying storage.
+        while self.buffer.buffer.lock().unwrap().len() >= self.threshold {
+            let chunk = {
+                let mut buffer = self.buffer.buffer.lock().unwrap();
+                buffer.split_to(self.threshold)
+            };
+            let size = chunk.len();
+
+            self.writer
+                .write_all(&chunk)
+                .await
+                .context(error::AsyncWriteSnafu)?;
+
+            self.flushed = true;
+            bytes_written += size as u64;
+        }
+
+        if all {
+            bytes_written += self.try_flush_all().await?;
+        }
+
+        self.bytes_written += bytes_written;
+
+        Ok(bytes_written)
+    }
+
+    async fn try_flush_all(&mut self) -> Result<u64> {
+        let remain = self.buffer.buffer.lock().unwrap().split();
+        let size = remain.len();
+
+        self.writer
+            .write_all(&remain)
+            .await
+            .context(error::AsyncWriteSnafu)?;
+        self.flushed = true;
+
+        self.bytes_written += size as u64;
```

Comment:
duplicated with line 123?

Reply:
However, It flushes the non-aligned(the remaining) chunk to the underlying writer.

Reply:
> However, It flushes the non-aligned(the remaining) chunk to the underlying writer.

I Moved it to `try_flush`

Reply:
Does the `bytes_written` accumulated twice?

Reply:
> Does the `bytes_written` accumulated twice?

Oh, you are right. And I fixed it.